pkg/utils/parallel/parallel.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parallel
+
+import (
+	"reflect"
+)
+
+// ParInput input for parallel execution
+type ParInput interface{}
+
+// ParOutput output for parallel execution
+type ParOutput interface{}
+
+type orderedParOutput struct {
+	ParOutput
+	index int
+}
+
+// ParExecProto parallel execute handler function on parInputs, with maximum concurrency as parallelism
+func ParExecProto(handler func(ParInput) ParOutput, parInputs []ParInput, parallelism int) []ParOutput {
+	outs := make(chan orderedParOutput)
+	pool := make(chan struct{}, parallelism)
```

Comment:
do we need to close the channal finally?

Reply:
Right, it would be more graceful to do so.